File: HarvestBench/claude.stale-2026-08-15-prompt-merge-62-63/mujs/translated/shim.c
/* C shim providing variadic public functions that the Rust cdylib cannot
 * define on stable. Each formats its arguments with vsnprintf (exactly like
 * the C original) then delegates to a Rust `*_impl` function. The public
 * linker symbols therefore match the original MuJS names.
 *
 * Compiled with -fexceptions/-funwind-tables so the Rust panic (used to model
 * longjmp) can unwind through these frames for the NORETURN error functions.
 */
#include <stdarg.h>
#include <stdio.h>
#include <string.h>

typedef struct js_State js_State;

/* Rust-side implementations (take a pre-formatted message). */
extern void rs_js_error(js_State *J, const char *msg);
extern void rs_js_evalerror(js_State *J, const char *msg);
extern void rs_js_rangeerror(js_State *J, const char *msg);
extern void rs_js_referenceerror(js_State *J, const char *msg);
extern void rs_js_syntaxerror(js_State *J, const char *msg);
extern void rs_js_typeerror(js_State *J, const char *msg);
extern void rs_js_urierror(js_State *J, const char *msg);

extern void rs_jsC_error(js_State *J, void *node, const char *msg);
extern void rs_jsP_error(js_State *J, const char *msg);
extern void rs_jsP_warning(js_State *J, const char *msg);
extern void rs_jsY_error(js_State *J, const char *msg);

#define FMT(buf) do { \
        va_list ap; va_start(ap, fmt); \
        vsnprintf((buf), sizeof(buf), fmt, ap); \
        va_end(ap); \
    } while (0)

void js_error(js_State *J, const char *fmt, ...)          { char b[256]; FMT(b); rs_js_error(J, b); }
void js_evalerror(js_State *J, const char *fmt, ...)      { char b[256]; FMT(b); rs_js_evalerror(J, b); }
void js_rangeerror(js_State *J, const char *fmt, ...)     { char b[256]; FMT(b); rs_js_rangeerror(J, b); }
void js_referenceerror(js_State *J, const char *fmt, ...) { char b[256]; FMT(b); rs_js_referenceerror(J, b); }
void js_syntaxerror(js_State *J, const char *fmt, ...)    { char b[256]; FMT(b); rs_js_syntaxerror(J, b); }
void js_typeerror(js_State *J, const char *fmt, ...)      { char b[256]; FMT(b); rs_js_typeerror(J, b); }
void js_urierror(js_State *J, const char *fmt, ...)       { char b[256]; FMT(b); rs_js_urierror(J, b); }

/* jsC_error/jsP_error/jsY_error format "file:line: " + message. The Rust side
 * builds the "file:line: " prefix; here we only format the user message part,
 * matching the C code which does: vsnprintf(msgbuf,...); snprintf(buf,"file:line: "); strcat. */
void jsC_error(js_State *J, void *node, const char *fmt, ...) { char b[256]; FMT(b); rs_jsC_error(J, node, b); }
void jsP_error_shim(js_State *J, const char *fmt, ...)   { char b[256]; FMT(b); rs_jsP_error(J, b); }
void jsP_warning_shim(js_State *J, const char *fmt, ...) { char b[256]; FMT(b); rs_jsP_warning(J, b); }
void jsY_error_shim(js_State *J, const char *fmt, ...)   { char b[256]; FMT(b); rs_jsY_error(J, b); }

File: HarvestBench/claude.stale-2026-08-15-prompt-merge-62-63/mujs/verified/src/shim.rs
//! Declarations of the C shim's variadic functions (defined in shim.c) plus
//! their Rust-side `rs_*` implementations. The shim formats varargs then calls
//! these. The public linker symbols (js_error, js_typeerror, ...) are exported
//! by the C shim, matching the original MuJS names.
#![allow(non_snake_case)]

use crate::types::*;
use std::os::raw::{c_char, c_int, c_void};

extern "C-unwind" {
    // Public variadic error functions (exported by shim.c with correct names).
    pub fn js_error(J: *mut js_State, fmt: *const c_char, ...);
    pub fn js_evalerror(J: *mut js_State, fmt: *const c_char, ...);
    pub fn js_rangeerror(J: *mut js_State, fmt: *const c_char, ...);
    pub fn js_referenceerror(J: *mut js_State, fmt: *const c_char, ...);
    pub fn js_syntaxerror(J: *mut js_State, fmt: *const c_char, ...);
    pub fn js_typeerror(J: *mut js_State, fmt: *const c_char, ...);
    pub fn js_urierror(J: *mut js_State, fmt: *const c_char, ...);

    // Internal variadic helpers (shim symbols with _shim suffix).
    pub fn jsC_error(J: *mut js_State, node: *mut c_void, fmt: *const c_char, ...);
    pub fn jsP_error_shim(J: *mut js_State, fmt: *const c_char, ...);
    pub fn jsP_warning_shim(J: *mut js_State, fmt: *const c_char, ...);
    pub fn jsY_error_shim(J: *mut js_State, fmt: *const c_char, ...);
}
